### DISCUSSION TITLE
A problem with selectAtom

### DISCUSSION THREAD

**Discussion by yekta** [pmndrs/jotai#1334] [Q&A] [answered]
I have an `itemsAtom` which is in the type of `Map<string,MyType>`. In a different component I want to be able to trigger re-renders when an object value belonging to a specific key on that map changes. This key happens to be the an `id` field belonging to the component I'm working with. I want to trigger updates specifically when the `quantity` field of a `MyType` object changes.

I'm doing the following in my component, but instead of component re-render being triggered when the `a?.quantity === b?.quantity` equality is false, the component gets re-rendered when `itemsAtom` changes.

```
const [item] = useAtom(
    useMemo(
      () =>
        selectAtom(
          itemsAtom,
          (i) => i.get(id),
          (a, b) => a?.quantity === b?.quantity
        ),
      []
    )
);

// for the sake of simplicity lets say MyType is the following
interface MyType {
    quantity: number;
}
```

I've been looking into this for the last hour and I still don't understand what's wrong with it. I double checked the docs, but feels like what I expect to happen should happen: "The selector function runs whenever the original atom changes; it updates the derived atom only if equalityFn reports that the derived value has changed".

Even if I do `(a, b) => true` for the equalityFn, the re-render still gets triggered. If I remove the block above, re-renders go away.

**Comment by dai-shi** [comment, marked as answer]
Do you update a map immutably?
```js
const newMap = new Map(oldMap)
newMap.set(key, newValue)
```

**Reply by yekta** [reply]
I think that's what I'm doing. The main atom gets modified somewhere else with new values via a write only atom. Simplified version: 

```
const addToItemsAtom = atom(
  null,
  (get, set, itemAddition) => {
    const items = get(itemsAtom);
    let newItems = new Map(items);
    const quantity =
      (items.get(itemAddition.id)?.quantity || 0) + (itemAddition.quantity || 1);
    newItems.set(itemAddition.id, { id: itemAddition.id, quantity });
    set(itemsAtom, newCart);
  }
);
```

Should I do something else when using maps?

**Reply by dai-shi** [reply]
It looks good then. You might want to try creating a minimal reproduction with codesandbox.

**Reply by yekta** [reply]
I wasn't expecting it to look good at all :D I'm quite new to this stuff, I assumed I made a mistake somewhere. I'll see if I can reproduce with minimal code.

**Reply by dai-shi** [reply]
> Even if I do `(a, b) => true` for the equalityFn, the re-render still gets triggered. If I remove the block above, re-renders go away.

This sounds really weird. Hope you can make a reproduction.

Oh, wait. Do you mean `console.log()` in the render body by "re-renders"?
Does it fix if you wrap `console.log()` with `useEffect`? If so, it's an expected behavior.

**Reply by yekta** [reply]
That's exactly what's happening/what I mean 😄 Is that a React noob error or something how did you know 😄?

More importantly, when that is the case, it means I'm not doing anything wrong on the Jotai side right, I'm using it properly?

**Reply by dai-shi** [reply]
Yes, you are using it correctly.

This behavior is confusing and many people have asked before.
It's pretty technical to explain, we re-evaluate the selector function in render phase and "bail out" if nothing is changed, which doesn't commit, thus no browser paint.

Jotai uses useReducer, and it is intentional useReducer behavior: https://twitter.com/dai_shi/status/1534170089981100032

**Reply by yekta** [reply]
Thanks a lot for the fast responses. Happy to see that I'm using it properly and didn't make a very simple mistake at least.
